### DISCUSSION TITLE
Task - Express - Mongoose - Local Library - CRUD Operation

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#46] [Tasks]
![library_website_-_mongoose_express](https://user-images.githubusercontent.com/60097560/173898417-eb51c696-1645-4674-8719-dc2ed9da327e.png)

**Comment by skagarwal5555** [comment]
//Create Schema
var booksSchema = new Schema({
	title:String,
	author:String,
	summary:String,
	ISBN:String,
	genre:String,
	url:String
});

//create model
var booksModel = mongoose.model('local_library',booksSchema);

//create model instance
var books_instance = new booksModel(
{
	title: "Learn Mern",
	author: "Akash Das",
	summary: "Details of Mern Stack",
	ISBN: "14525",
	genre: "Thriller",
	url: "Github\akash-coded"
});

//Save record
books_instance.save(function(err,result) {
if(err) return handleError(err);
console.log(result);
});

//create another instance
var books_instance = new booksModel(
{
	title: "Learn MEAN",
	author: "Sourav Das",
	summary: "Details of MEAN Stack",
	ISBN: "1452545",
	genre: "Thriller",
	url: "Github\akash-coded\MEAN"
});

//save record
books_instance.save(function(err,result) {
if(err) return handleError(err);
console.log(result);
});

//find record
booksModel.find({ author: 'Akash Das'},"title author genre", function (err, docs) {
    if (err){
        console.log(err);
    }
    else{
        console.log(docs);
    }
});

//update record
booksModel.updateOne({author:: 'Akash Das'}, 
    {summary:"New changes for MERN Stack included"}, function (err, docs) {
    if (err){
        console.log(err)
    }
    else{
        console.log(docs);
    }
});

//delete record
booksModel.deleteOne({ author:: 'Sourav Das' }).then(function(){
    console.log("Data deleted");
}).catch(function(error){
    console.log(error);
});``

**Comment by MRamitkv** [comment]
``const express = require("express");
const app = express();
const bodyParser = require('body-parser');
const port = 8000;
var mongoose = require('mongoose');
//const { isModuleNamespaceObject } = require("util/types");
// Defining the endpoint

app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json());
app.use(bodyParser.raw());

//Set up default mongoose connection
var mongoDB = 'mongodb://127.0.0.1/local_library';
mongoose.connect(mongoDB, {useNewUrlParser: true, useUnifiedTopology: true});

//Get the default connection
var db = mongoose.connection;
db.on('error', console.error.bind(console, 'MongoDB connection error:'));


var bookSchema = new mongoose.Schema({
    title:String,
    author: [{
        first_name: String, 
        family_name: String,
        date_of_birth: Date,
        date_of_death : Date,
        name: String,
        lifespan: String,
        url: String
    }],
    summary: String,
    ISBN: String,
    genre: [{
        name: String,
        url: String
    }],
    url: String
})
var bookInstanceSchema = new mongoose.Schema({
  book: bookSchema,
  imprint: String,
  status: {
    type: String,
    enum: ["pending", "issued","returned","available"],
    default: "available"
  },
  due_back: Date
})

var bookInstanceModel = new mongoose.model("bookinstance", bookInstanceSchema);

app.post("/book", (req, res) => {
    console.log(JSON.stringify(req.body));
    var bookinstance = new bookInstanceModel(req.body);
    bookinstance.save(function(err, instance){
        console.log(JSON.stringify(err))
        res.send("Book details inserted "+JSON.stringify(req.headers)+" -----------8888------ "+JSON.stringify(instance._id));
    })
  });
  
  app.post("/book/find", (req, res) => {
    console.log(JSON.stringify(req.body));
    var bookinstance =  bookInstanceModel.find(req.body, function(err, book){
        console.log(JSON.stringify(book))
        if(book)
        res.send(book);
        else
        res.send("not found")
    });

    
  });
  




// Starting the server
app.listen(port, () => {
    console.log(`HelloWorld app listening on port ${port}!`);
  });``

**Comment by Subbu1010** [comment]
var bookSchema =  new Schema({
title: { type: String, required : true},
author: { type: String, required : true},
summary: { type: String, required : false},
ISBN:{ type: Number, required : true},
url:{ type: String, required : true},
})

var bookModel = mongoose.model('book',bookSchema);

var modelInstance = new 

bookModel.save(function(error, output){
            if(!error){console.log(output)};
    })

bookModel.findById(function(error, output){
            if(!error){console.log(output)};
    })

bookModel.findByIdAndDelete(function(error, deletedId){
            if(!error){console.log(deletedId)};
    })

**Comment by santu1667** [comment]
Import and connect to MongoDB from app.js

var mongoose = require('mongoose');
var mongoDb = "mongoDB://127.0.01/book_Collection";
mongoose.connect(mongoDb,{useNewUrlParser:true, useUnifiedTopology:true});
var db =mongoose.connection();


Create Schema 

var bookSchema = new Schmea({
							title :
								{
								type:String,
								required:[true,'Title cannot be Empty']
								},
							authorName:
								{
								type:String,
								required:[true,'Author cannot be Empty']
								},
							Summary:
								{
								type:String
								},
							genreName:
								{
								value:"Fantasy"
								type:String,
								},
							iSBN:
								{
								type:String
								}
							});
//create Model
var bookModel = mongoose.model ("bookCollection",bookSchema);
//create instance
var book_instance = new bookModel({schema});
book_instance.save(function(err)
						{// error handling
						})
console.log('Title of the book was  ' + book_instance.title)
						
//Read Instance
var query = bookCollection.find({"title:"Harry Potter"})
query.exec(function(error, results){
if(err){}else console.log(results)}).


//Delete Instance

var query = bookCollection.find({"title":"Harry Potter"}).remove;
query.exec(function(error, results){
if(err){}else console.log(results)}).

**Comment by surajsukumar** [comment]
var bookSchema = new mongoose.Schema({
    title: String,
    author: String,
    summary: String,
    ISBN: String,
    genre: {
      name: String,
      url: String,
    },
    url: String,
  });

  var bookModel = mongoose.model("Book", bookSchema);
  var bookInstance = new bookModel({
    title: "MongoDB Basics",
    author: "ABC",
    summary: "This book is a basic book of MongoDB",
    ISBN: "1334543423",
    genre: {
      name: "Technology",
      url: "www.mongoadfsd.com",
    },
    url: "www.mongodbbasics.com",
  });

  //Save
  bookInstance.save(function (err) {
    if (err) {
      console.log(err);
    }
  });

  //Find
  bookInstance.find({ author: "ABC" }, function (err, result) {
    if (err) {
      console.log(err);
    } else {
      console.log(result);
    }
  });

  //Update
  bookInstance.updateOne(
    { author: "ABC" },
    { summary: "This is a book for teaching basics of MongoDB" },
    function (err, result) {
      if (err) {
        console.log(err);
      } else {
        console.log(result);
      }
    }
  );

**Comment by sanath8se** [comment]
```
var mongoose = require('mongoose');

mongoose.connect("mongodb://http://localhost:27017/local_library");

var books = new mongoose.Schema({
  title: { type: String, required: true },
  _authorIds: [mongoose.Schema.Types.ObjectId],
  summary: { type: String },
  isbn: { type: String, required: true },
  genre: [mongoose.Schema.Types.ObjectId],
  url: { type: String },
});
var authors = new mongoose.Schema({
  first_name: { type: String, required: true },
  _authorId: {
    type: mongoose.Schema.Types.ObjectId,
    index: true,
    required: true,
    auto: true
  },
  family_name: { type: String },
  dob: { type: String, required: true },
  url: { type: String },
});
// var genres = new mongoose.Schema({
//   name: { type: String, required: true },
//   __genreId: {
//     type: mongoose.Schema.Types._genreId,
//     index: true,
//     required: true,
//     auto: true
//   },
//   url: { type: String }
// });

var authorModel = mongoose.model('authors', authors);
var bookModel = mongoose.model('books', books);

var john = new authorModel({ first_name: 'John', family_name: 'Doe', dob: '02/19/2000' });

john.save(function (err) {
  if (err) return createError(500);

  var book = new bookModel({
    title: "be yourself",
    _authorIds: john._authorId,
    summary: "about being yourself",
    isbn: "er3423sa",
    genre: "non-fiction"
  });

  book.save(function (err) {
    if (err) return createError(500);

  });
});

//find record
bookModel.find({ author: 'John Doe' }, "title summary", function (err, docs) {
  if (err) {
    console.log(err);
  }
  else {
    console.log(docs);
  }
});

//update record
booksModel.updateOne({ author: 'John Doe' },
  { summary: "do more" }, function (err, docs) {
    if (err) {
      console.log(err)
    }
    else {
      console.log(docs);
    }
  });

//delete record
booksModel.deleteOne({ author: 'John Jacob' }, function () {

});
```

**Comment by CostaPasta** [comment]
```
var bookSchema = Schema({
    title: String,
    Author: {type: Schema.Types.ObjectId, ref: 'Author'},
    summary: String,
    ISBN: String,
    url: String
});

var authorSchema = Schema({
    name: String,
    lifespan: String,
    url: String,
    books: [{type: Schema.Types.ObjectId, ref: 'Book'}],
    genre: [{type: Schema.Types.ObjectId, ref: 'Genre'}]
});

var genreSchema = Schema({
    name: String,
    url: String
});

var bookInstanceSchema = Schema({
    book: bookSchema,
    imprint: String,
    status: Enumerator,
    due_back: Date,
    url: String
})
```

**Comment by sanath8se** [comment]
```
var bookSchema = Schema({
  title: String,
  author: { type: Schema.Types.ObjectId, ref: 'Author' },
  genre: { type: Schema.Types.ObjectId, ref: 'Genre' },
  summary: String,
  ISBN: String,
  url: String
});

var genres = new mongoose.Schema({
  name: { type: String, required: true },
  url: { type: String }
});

var book = mongoose.model('Book', bookSchema);
var genre = mongoose.model('Genre', genres);

book.findOne({ title: "The Jungle Book" })
  .populate('author', 'genre')
  .exec(function (err, book) {
    if (err) return handleError(err);
    console.log(book.author.name);
  });
```

**Comment by Neethumar** [comment]
```javascript
var mongoose = require('mongoose')
 
//Schemas
var bookSchema = new mongoose.Schema({
   	title     : String,
   	author  : [{ type: mongoose.Schema.Types.ObjectId, ref: 'Author' }],
   	summary : String,
   	ISBN   : String,
	genre:[{ type: mongoose.Schema.Types.ObjectId, ref: 'Genre' }].
	url:String
});

var authorSchema =new mongoose.Schema({
   	first_name: String,
   	family_name: String,
   	date_of_birth: Date,
	date_of_death:Date,
   	name: String,
   	lifespan: String,
   	url:String
});

var genreSchema=new mongoose.Schema({
	name:String;
	url:String
})

var bookinstanceSchema=new mongoose.Schema({
	book:[{ type: mongoose.Schema.Types.ObjectId, ref: 'Book' }],
	imprint:String.
	status:enum,
	due_back:Date,
	url:String
})

//Models
var Book  = mongoose.model('Book', bookSchema);
var Genre  = mongoose.model('Genre', genreSchema);
var Author  = mongoose.model('Author', authorSchema);
var BookInstance = mongoose.model('BookInstance', bookinstanceSchema);
```

**Comment by shubhamk-coder** [comment]
## Code
```javascript
let mongoose = require("mongoose");

// Mongoose Schemas
let books = new mongoose.Schema({
  title: String,
  author: [{ type: mongoose.Schema.Types.ObjectId, ref: "Author" }],
  summary: String,
  ISBN: String,
  genre: [{ type: mongoose.Schema.Types.ObjectId, ref: "Genre" }],
  url: String,
});

let authors = new mongoose.Schema({
  first_name: String,
  family_name: String,
  date_of_birth: Date,
  date_of_death: Date,
  name: String,
  lifespan: String,
  url: String,
});

let genres = new mongoose.Schema({
  name: String,
  url: String,
});

let bookInstance = new mongoose.Schema({
  book: [{ type: mongoose.Schema.Types.ObjectId, ref: "Book" }],
  imprint: String,
  status: String,
  due_back: Date,
  url: String,
});

// Mongoose Models
let Book = mongoose.model("Book", books);
let Genre = mongoose.model("Genre", genres);
let Author = mongoose.model("Author", authors);
let BookInstance = mongoose.model("BookInstance", bookInstance);
```

**Comment by KNiranjani** [comment]
**//Creating Mongoose Schema**
```js
var bookSchema = new mongoose.Schema({
   	title     : String,
   	author  : [{ type: mongoose.Schema.Types.ObjectId, ref: 'Author' }],
   	summary : String,
   	ISBN   : String,
	genre:[{ type: mongoose.Schema.Types.ObjectId, ref: 'Genre' }],
	url:String});

var authorSchema =new mongoose.Schema({
   	first_name: String,
   	family_name: String,
   	date_of_birth: Date,
	date_of_death:Date,
   	name: String,
   	lifespan: String,
   	url:String});

var genreSchema=new mongoose.Schema({
	name:String;
	url:String});

var bookinstanceSchema=new mongoose.Schema({
	book:[{ type: mongoose.Schema.Types.ObjectId, ref: 'Book' }],
	imprint:String,
	status:enum,
	due_back:Date,
	url:String});

```
**//Creating Mongoose Models**
```js
var Book  = mongoose.model('Book', bookSchema);
var Genre  = mongoose.model('Genre', genreSchema);
var Author  = mongoose.model('Author', authorSchema);
var BookInstance = mongoose.model('BookInstance', bookinstanceSchema);
```

**//find record**
```js
booksModel.find({ author: 'Niran jani'},"title author genre", function (err, docs) {
if (err){
console.log(err);
}
else{
console.log(docs);
}
});
```

**//update record**
```js
booksModel.updateOne({author:: 'Niran jani'},
{summary:"New Discussion is updated"}, function (err, docs) {
if (err){
console.log(err)
}
else{
console.log(docs);
}
});
```

**//delete record**
```js
booksModel.deleteOne({ author:: 'Prithik Singh' }).then(function(){
console.log("Record has been deleted");
}).catch(function(error){
console.log(error);
});
```
